server/src/main/java/org/elasticsearch/index/mapper/TextFieldExactQuery.java

```diff
@@ -0,0 +1,153 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.LeafFieldData;
+import org.elasticsearch.script.field.DocValuesScriptFieldFactory;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Objects;
+
+/**
+ * Find documents with text fields that exactly match the input
+ */
+public class TextFieldExactQuery extends Query {
+
+    private final String field;
+    private final String value;
+    private final Query conjunction;
+    private final IndexFieldData<?> fieldData;
+
+    public TextFieldExactQuery(MappedFieldType fieldType, IndexFieldData<?> fieldData, String input) {
+        this.field = fieldType.name();
+        this.value = input;
+        this.fieldData = fieldData;
+        this.conjunction = buildConjunction(
+            fieldType.name(),
+            fieldType.getTextSearchInfo().searchAnalyzer().tokenStream(fieldType.name(), input)
+        );
+    }
+
+    private TextFieldExactQuery(String field, String value, Query conjunction, IndexFieldData<?> fieldData) {
+        this.field = field;
+        this.value = value;
+        this.conjunction = conjunction;
+        this.fieldData = fieldData;
+    }
+
+    private static Query buildConjunction(String field, TokenStream ts) {
+        BooleanQuery.Builder bq = new BooleanQuery.Builder();
+        CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+        int count = 0;
+        try {
+            ts.reset();
+            while (ts.incrementToken() && count++ < 1000) { // limit the size of the approximation
+                bq.add(new TermQuery(new Term(field, termAtt.toString())), BooleanClause.Occur.FILTER);
+            }
+            ts.end();
```

Comment:
I don't think it's legal to call `end` if the token stream may not be fully consumed. We may need to consume the token stream without doing anything with the tokens like `LimitTokenCountTokenFilter` does.